File: Data Structures & Algorithms/longest-consecutive-sequence/submission-6.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        dedup_nums = set(nums)
        if not dedup_nums:
            return 0

        best = 1

        for num in dedup_nums:
            if num-1 in dedup_nums:
                continue
            
            i = 1
            while num + i in dedup_nums:
                best = max(best, i+1)
                i += 1
            
        return best

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-6.go
func lengthOfLIS(nums []int) int {
	mem := map[int]int{}
	
    var rec func(int) int
	rec = func(i int) int {
		if val, ok := mem[i]; ok {
			return val
		}

		best := 1
		for j := i+1; j < len(nums); j++ {
			if nums[i] < nums[j] {
				best = max(best, rec(j) + 1)
			}
		}

		mem[i] = best
		return best
	}

	best := 1
	for i := range nums {
		best = max(best, rec(i))
	}

	return best
}
